File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        ml = ListNode(0)
        tr = ml

        while list1 and list2:
            if list1.val < list2.val:
                ml.next = list1
                list1 = list1.next
            else:
                ml.next = list2
                list2 = list2.next
            ml = ml.next

        ml.next = list1 if list1 else list2

        return tr.next
        

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-2.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) {
            return nullptr;
        }

        auto node_comparator = [](const ListNode* a, const ListNode* b) {
            return a -> val >= b -> val;
        };
        priority_queue<ListNode*, vector<ListNode*>, decltype(node_comparator)> pq;

        for (auto linked_list : lists) {
            if (linked_list) {
                pq.push(linked_list);
            }
        }

        // now we have the first nodes of all the lists.
        ListNode* result = new ListNode();
        ListNode* ptr = result;
        while (!pq.empty()) {
            auto top_element = pq.top();
            pq.pop();
            ptr -> next = top_element;
            ptr = ptr -> next;
            if (top_element -> next) {
                pq.push(top_element -> next);
            }
        }

        return result -> next;
    }
};
